hw/ip/otbn/dv/tracer/cpp/log_trace_listener.cc

```diff
@@ -0,0 +1,63 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cassert>
+#include <iomanip>
+#include <ios>
+#include <sstream>
+#include <string>
+
+#include "log_trace_listener.h"
+
+LogTraceListener::LogTraceListener(const std::string &log_filename)
+    : trace_log(log_filename, std::fstream::out) {
+  if (!trace_log.is_open()) {
+    std::ostringstream oss;
+    oss << "Could not open log file: " << log_filename;
+    throw std::runtime_error(oss.str());
+  }
+}
+
+void LogTraceListener::AcceptTraceString(const std::string &trace,
+                                         unsigned int cycle_count) {
+  assert(trace_log.is_open());
+
+  // Split the trace up into a vector of strings, one per line
+  auto trace_lines = SplitTraceLines(trace);
+  bool first_line = true;
+
+  // Write out the lines from the trace
+  for (auto &line : trace_lines) {
+    if (first_line) {
+      // Special handling for the first line
+      assert(line.size() > 1);
+
+      // It is expected the first line of any trace output is an 'E' or 'S' line
+      // (instruction execute or instruction stall)
+      bool is_e_or_s_line = line[0] == 'E' || line[0] == 'S';
+
+      // Output the beginning of the first line adding a cycle count. If this
+      // isn't an 'E' or 'S' line use a special '!' to indicate the lack of one.
```

Comment:
When does the '!' case happen?

Reply:
Re-worded it to hopefully make this more obvious.

It's where we have a cycle where OTBN does things (namely writes to registers or memory) without an associated instruction. This shouldn't actually happen so if it does wanted the '!' to highlight the oddity. In verification this should result in a mismatch because the model won't emit such state changes without them being part of an instruction.

(We actually get a couple of these '!' lines to start with, something around initial state and reset state not being quite as they should, need to fix that but that bug revealed the need for something like this).

Reply:
Ah, I see. Right, that makes sense. Could you add the '!' case to the big comment at the top of `otbn_tracer.sv`? I'd actually consider something more like "Trace output is generated as a series of records, at most one per cycle. Each record starts with an 'E', 'S' or '!' line, representing an instruction executing, an instruction stalling, and an exceptional condition, respectively. Following that line are zero or more lines starting with '<' or '>', for state reads and writes, respectively."

Reply:
So the '!' lines aren't part of the raw format and hence have nothing to do with `otbn_tracer.sv`. They're introduced by the log output mostly because of its attempt to indent all of the 'activity' lines below the 'execution' line that tells you about the instruction. This breaks where we have a cycle with activity but no execution (you either have something like a '>' line with a cycle count that doesn't get indented that looks odd, or you realise this and just dont have the un-indented line for that record but then you don't know what cycle count is associated with it and you may confuse it with activity from another cycle).

For the ISS compare for instance the '!' lines don't exist, you could view them purely as part of the presentation layer of the logging (much in the way the indentation and the added cycle counts are).

Reply:
Oh, I see! So I should have written something like this? "Trace output is generated as a series of records, at most one per cycle. A record may start with an 'E' or 'S' line, representing an instruction executing or stalling, respectively. Following that line are zero or more lines starting with '<' or '>', for state reads and writes, respectively. No record is empty (there will always be at least one line)."

Reply:
Almost, you can also get 'R' and 'W' lines (for memory reads and writes), plus I want to leave open the idea that other categories will crop up, hence referring the reader to look at the prefix strings declared at the top of the module (e.g. we may want to add an 'L' for loop related stuff and potentially extra categories around CFI functionality).

I should maybe add comments to the prefix declarations to explain what they're used for along with the format that use.

Reply:
Wait, what? I thought that we had a record per cycle, which contained various things happening (register state; memory traffic; loop state). Have I got this format completely wrong?

Reply:
Yes exactly that. I'm using the following terminology:

Record - The bunch of string data a single call to `accept_otbn_trace_string` receives
Line - One of the lines within the record

You receive 0 or 1 records per cycle (which cycle is given as a seperate argument to `accept_otbn_trace_string`).

If a record has an 'E' or 'S' line it will be the first line of a record.

Other records could have a variety of prefixes, '>', '<', 'R', 'W' currently but more may be added

Here is an example of a record that contains a 'W' line:

```
E PC: 0x00000158, insn: 0x01acd08b
< w20: 0x78fccc06_2228e9d6_89c9b54f_887cf14e_c79af825_69be57d4_fecd21a1_b9dd0141
< x25: 0x00000020
< x26: 0x00000014
> x26: 0x00000015
W 0x[00000020]: 0x78fccc06_2228e9d6_89c9b54f_887cf14e_c79af825_69be57d4_fecd21a1_b9dd0141
```

Reply:
Cool, makes sense.